stage4.py

```diff
@@ -1,48 +1,46 @@
 from card import Card
+from operator import itemgetter
 import random
 
-
+# Adding cards to the deck
```

Comment:
As we are going to add more and more logic into a file it might become very long to read than new good developers (הטובים מבין המפתחים החדשים) understand they need to find a solution for that and start to give "titles" to part of the codes (like that line) or start separate it with a lot of '-------------' as a comment. It's a really good idea because solve if the issue but isn't the best practice. 

Best practice - תיאור לדרך עיצובית הטובה והמקובלת ביותר 

**Why comments are not the best for splitting the code into smaller parts:** 
1. it's hard to navigate or search you can (Ctrl+f) but it search in the "titles" and in the code. 
2. very fast you can get to 1000 lines of code in the same file (with a comment it even rush that) 
3. the code changes very fast while the developers tend not to update the comments. 
4. code isn't reusable 
5. comments don't give you good overview on the flow (for example we would like to have a table of contents on all the titles)

Reply:
**What is good practice?** "Function 💃" (and sometimes even file split, but let's wait with it) 

instead of having 
```
# get a list of a student from csv 
... a lot of code ... 
# Search student from HIT that failed on two test 
...code...
# print all those students 
... code ...
```

You can have 


```
CSV_FILE = 'c:/data/students.csv'
def main():
  students = load_student_from_csv(CSV_FILE)
  failed_students = get_failed_students(students)
  print_students(failed_students)
```

In that way: 
1. you can understand the flow much better 
2. easy to debug
3. to change the code or one specific logic should be easier because you need to read a smaller scope of code
4. solve all the issues from the last comment.

Reply:
The next question that it is raised, how often I need to use functions? (every 5 lines? 10? ,20? 50?)

The answer is very hard, and really depends on what you think would be the best way to read your code. 
if you feel you want to add more "titles" into your code -> split it into more functions 
if you feel the name is the same like the code -> don't use a function 
(
example: 
```
#not good reason for having a function as a "title" 
def set_student_ago(student, age)
  student.age = age
```
)

Reply:
מצטער על החפירה, מנסה לעשות את זה פעם ראשונה בכתב
אפשר לעשות זום על זה אם תרצה :)